FLT/HaarMeasure/HaarChar/FiniteDimensional.lean

```diff
@@ -1,17 +1,207 @@
+/-
+Copyright (c) 2025 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard, Norbert Voelker
+-/
 import FLT.HaarMeasure.HaarChar.Ring
 import FLT.Mathlib.Topology.Algebra.Module.Equiv
 import FLT.Mathlib.LinearAlgebra.Determinant
 import Mathlib.Topology.Algebra.Module.ModuleTopology
+import FLT.Mathlib.Topology.Algebra.Module.Equiv
 
 namespace MeasureTheory
 
 open Measure
 
+section HaarMeasureScaling
+
+/-!
+
+# Haar character scaling under linear automorphisms
+
+This section proves that a linear isomorphism on a finite-dimensional vector space
+`V` over a locally compact field `F` scales Haar measure by its determinant. The
+proof assumes `[SecondCountableTopology F]`.
+
+We proceed by:
+1. Proving the scaling property for diagonal matrices.
+2. Showing that transvections preserve Haar measure.
+3. Extending to arbitrary matrices using induction on diagonal/transvection decomposition.
+4. Proving the result for automorphisms on `(ι → F)`
+5. Deducing the general result for `V`
+
+## Main result:
+
+* `addEquivAddHaarChar_eq_ringHaarChar_det`
+
+## TODO
+
+* Review `[SecondCountableTopology F]` assumption in the future
+
+-/
+
+open Matrix Measure
 
 variable {F : Type*} [Field F] [MeasurableSpace F] [TopologicalSpace F] [BorelSpace F]
   [IsTopologicalRing F] [LocallyCompactSpace F]
+variable {ι : Type*} [Fintype ι] [DecidableEq ι]
+
+/-- A diagonal matrix scales addHaar with its determinant
+-/
+lemma addEquivAddHaarChar_eq_ringHaarChar_det_diagonal
+    [BorelSpace (ι → F)]
+    (ρ : (ι → F) ≃L[F] (ι → F)) {D : ι → F}
+    (h : ρ.toLinearMap.toMatrix' = Matrix.diagonal D) :
+  addEquivAddHaarChar ρ.toContinuousAddEquiv
+    = ringHaarChar ρ.toLinearEquiv.det := by
+  -- 1) determinant computations
+  let f := ρ.toLinearMap
+  have f_eq : f = Matrix.toLin' (Matrix.diagonal D) := by
+    rw [← h, Matrix.toLin'_toMatrix']
+  have f_det: f.det = (∏ i, D i) := by
+    simp only [f_eq, LinearMap.det_toLin', det_diagonal]
+  have det_ne : f.det ≠ 0 := by
+  -- we know det(ρ) is a unit, so nonzero
+    have := LinearEquiv.coe_det ρ.toLinearEquiv
+    simpa [this] using Units.ne_zero (ρ.toLinearEquiv.det)
+  have prod_ne : (∏ i, D i) ≠ 0 := by
+    -- det(diagonal D) = ∏ i, D i
+    simp [← f_det, det_ne]
+  have D_ne : ∀ i, D i ≠ 0 :=
+    fun i => (Finset.prod_ne_zero_iff.mp prod_ne) i (by simp)
+  have det_equiv :
+      (↑(ρ.toLinearEquiv.det) : F) = ∏ i, D i := by
+      have: ↑(ρ.toLinearEquiv.det) = f.det := LinearEquiv.coe_det ρ.toLinearEquiv
+      simp [this, f_det]
```

Comment:
```suggestion
  have det_equiv : (↑(ρ.toLinearEquiv.det) : F) = ∏ i, D i := by
    have: ↑(ρ.toLinearEquiv.det) = f.det := LinearEquiv.coe_det ρ.toLinearEquiv
    simp [this, f_det]
```
Prefer 1 line if possible. And indent 2 for proofs.

Reply:
done